src/Context/ContextHandlerTrait.php

```diff
@@ -30,61 +32,174 @@
+          throw new RulesEvaluationException("The context for $name is NULL, but the context $name in "
             . $plugin->getPluginId() . ' requires a value.');
         }
+      }
+    }
+  }
 
-        $context = $plugin->getContext($name);
-        $new_context = Context::createFromContext($context, $this->configuration['context_values'][$name]);
-        $plugin->setContext($name, $new_context);
+  /**
+   * Prepares plugin context based upon the set context configuration.
+   *
+   * The configuration is applied as far as possible without having execution
+   * time data. That means, the configured context values are set and context is
+   * refined while leveraging the definitions of selected data.
+   *
+   * @param \Drupal\Core\Plugin\ContextAwarePluginInterface $plugin
+   *   The plugin that is prepared.
+   * @param \Drupal\rules\Engine\ExecutionMetadataStateInterface $metadata_state
+   *   The metadata state, prepared for the current expression.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\ContextException
+   *   Thrown if the plugin tries to access some not-defined context. As this is
+   *   a developer error, this should not be caught.
+   *
+   * @see ::prepareContext()
+   */
+  protected function prepareContextWithMetadata(CoreContextAwarePluginInterface $plugin, ExecutionMetadataStateInterface $metadata_state) {
+    if (isset($this->configuration['context_values'])) {
+      foreach ($this->configuration['context_values'] as $name => $value) {
+        $plugin->setContextValue($name, $value);
       }
-      elseif ($definition->isRequired()) {
-        throw new RulesEvaluationException("Required context $name is missing for plugin "
-          . $plugin->getPluginId() . '.');
+    }
+
+    $selected_data = [];
+    // Collected the definitions of selected data for refining context
+    // definitions.
+    if (isset($this->configuration['context_mapping'])) {
+      // If no state is available, we need to fetch at least the definitions of
+      // selected data for refining context.
+      foreach ($this->configuration['context_mapping'] as $name => $selector) {
+        try {
+          $selected_data[$name] = $this->getMappedDefinition($name, $metadata_state);
+        }
+        catch (RulesIntegrityException $e) {
+          // Ignore invalid data selectors here, such that context gets refined
+          // as far as possible still and can be respected by the UI when fixing
+          // broken selectors.
+        }
       }
     }
+
+    if ($plugin instanceof ContextAwarePluginInterface) {
+      // Getting context values may lead to undocumented exceptions if context
+      // is not set right now. So catch those exceptions.
+      // @todo: Remove ones https://www.drupal.org/node/2677162 got fixed.
+      try {
+        $plugin->refineContextDefinitions($selected_data);
+      }
+      catch (ContextException $e) {
+        if (strpos($e->getMessage(), 'context is required') === FALSE) {
```

Comment:
what is going on here? you check for some kind of exception? so you only throw required context exceptions and eat other silently? Can you add a comment?